Extract Variables from JobAccountant Profiling

d-ylee opened the pull request:
Fixes #12007 

#### Status
In development

#### Description
Adds another context manager to dump variables to a JSON file for testing changes to containerized MariaDB. An extension of @amaltaro 's work in #12008 .

#### Is it backward compatible (if not, which system it affects?)
MAYBE

#### Related PRs
#12008 

#### External dependencies / deployment changes
No

amaltaro committed:
Debug/profile JobAccountant

d-ylee committed:
Saving variables after profiling in AccountantWorker

Initial implementation for WMBS

amaltaro commented:
@d-ylee can you please rebase this PR? It brings in many other changes as well.

d-ylee commented:
@amaltaro Yea I just noticed. It should be fixed now, I think.

amaltaro commented on src/python/WMComponent/JobAccountant/AccountantWorker.py:
```diff
@@ -244,7 +251,17 @@ def __call__(self, parameters):
+            filepath = pathlib.Path(self.compDir, 'WMBS_unmerged.json')
+            if not (filepath.exists() and filepath.stat().st_size > 0):
+                f = stack.enter_context(open(filepath, 'w'))
+                content = {'wmbsFilesToBuild': self.wmbsFilesToBuild,
```
Dennis, I would suggest saving 2 different files - and probably name them after the variable name. IMO it will be more clear and avoid any mistakes/confusion

amaltaro reviewed (commented):
Thanks Dennis. I left one comment along the code.

In addition to that, I am not really familiar to this ExitStack class, but if it does the job, we are good. Don't worry if the code looks ugly or not performant, we will only apply it to the backfill agent and we can easily remove the patch later on.

d-ylee committed:
Added ctx to other targeted functions, separate files for variables

d-ylee committed:
Profile the 4 functions in handleWMBS, updated CodeTimer

Created CodeTimer2 to save profiling information, variable content to a
json file

d-ylee committed:
pickle instead of json

amaltaro commented:
@d-ylee I believe this debugging PR is no longer relevant, so I will just go ahead and close it out. Please reopen and/or let me know if there is anything that I missed. Thanks!
